net: support multiple network interfaces

slp opened the pull request:
So far, we were assuming the microVM will have a single network interface. But we have at least one new usecase (namely, running Android), that requires having multiple interfaces.

Refactor the API to support adding mulitple interfaces, while we keep compatibility with the old API.

While there, also allow the user to set up the network device flags.

slp commented on include/libkrun.h:
```diff
@@ -265,7 +265,63 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+int32_t krun_add_net_gvproxy(uint32_t ctx_id,
+                             char *c_path,
+                             uint8_t *const c_mac,
+                             uint32_t flags);
```
This poses an interesting question. While it's true that QEMU uses generic terms like `socket` for referring to the backend of a `chardev`, applications such as `passt` or `gvproxy` (in vfkit mode) behave in significantly different ways. Does `vnet` behave in exactly the same was as `gvproxy` in vfkit mode? If so, we could perhaps rename this function (and the backend) as `*_vfkit`.

nirs replied:
The same datagram socket works with vfkit and krunkit and the vmnet framework. Any frame sent to the socket is written to vment, and any frame read from vmnet is written to the socket.

vmnet-helper is basically a stupid proxy between the socket and the vment framework:
- https://github.com/nirs/vmnet-helper/blob/2d26ed25c87fbc0ded832211a04e2f4c13949313/helper.c#L719
- https://github.com/nirs/vmnet-helper/blob/2d26ed25c87fbc0ded832211a04e2f4c13949313/helper.c#L632

vment-helper also work with qemu using `-netdev drgram`:
https://github.com/nirs/vmnet-helper/blob/2d26ed25c87fbc0ded832211a04e2f4c13949313/vmnet/vm.py#L192

There is one little hack vment-helper does to enable vfkit and krunkit with unix socket, it drops the invalid VFKT packet used as a "handshake":
https://github.com/nirs/vmnet-helper/blob/2d26ed25c87fbc0ded832211a04e2f4c13949313/helper.c#L459

The handshake packet is not really needed and we can drop it from gvproxy and vfkit.

Using krun_add_net_vfkit() is also not great since vfkit is just a wrapper for the virtualization framework.

So I would go with `krun_add_net_unixgram()` or `krun_add_net_socket()`, document the semantics (proxy frames between the guest and the network stack) and mention gvproxy and vment-helper as programs that can work with this socket.

slp replied:
We still need to keep compatibility with vfkit, so instead of proposing a change to `gvproxy` I'm going to rename the function to `*_unixgram` and add a non-virtio flag to indicate we need to send the VFKIT magic. @nirs Does this sound reasonable to you?

cfergeau replied:
https://github.com/cfergeau/gvisor-tap-vsock/commit/a13a3dca8c380c7f12325ea78c4150227c69e29c This is what I tried at the time.

nirs replied:
I think supporting both fd and unix socket in libkrun is the way to keep the library easy to use. At the backend level we already use a fd, obtained from the unix socket. We can add 2 "constructors" one for unix socket and one for fd to support both.

slp replied:
Actually, while working on this PR, I've considered removing the socket initialization logic from `Gvproxy` and forcing users to do that work for us. But that would be a user-breaking change we can't do in v1.x.x, and keeping both mechanisms for compatibility seems like a waste of code. I'm inclined to keep the behavior of `krun_set_gvproxy_path`, with the rename and making the magic optional, and do something more generic in v2.x.x.

mtjhrc replied:
> I think supporting both fd and unix socket in libkrun is the way to keep the library easy to use. 
Yes this is also a good argument, so I'm fine with this either way.

I was just throwing it out there that supporting only the file descriptor case would remove the need for conditionally sending the "VFKT" magic by libkrun (by making the library user do it). Is it a good idea? I'm not sure.

cfergeau replied:
> > [cfergeau/gvisor-tap-vsock@a13a3dc](https://github.com/cfergeau/gvisor-tap-vsock/commit/a13a3dca8c380c7f12325ea78c4150227c69e29c) This is what I tried at the time.
> 
> I think the issue is reading with empty buffer:
> 
> ```
> syscall.Recvfrom(int(fd), []byte{}, syscall.MSG_PEEK|syscall.MSG_TRUNC)
> ```
> 

I tried it again, and this seems to work this time https://github.com/cfergeau/gvisor-tap-vsock/commit/8aace499d80ced197a90dd7f4b5fbfea86331bb0 
Maybe I missed the retry loop the first time.
I filed https://github.com/containers/gvisor-tap-vsock/issues/520 so that we get this into shape.

slp commented on include/libkrun.h:
```diff
@@ -265,7 +265,63 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+ *
+ * Arguments:
+ *  "ctx_id"  - the configuration context ID.
+ *  "fd"      - a file descriptor to communicate with passt.
```
The nature of the `fd` is defined by the userspace application we're connecting to (in this case, passt). This function doesn't aim to be generic, as realistically this isn't possible as exposed in the previous comment.

slp commented on include/libkrun.h:
```diff
@@ -265,7 +265,63 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+#define NET_HOST_UFO 1 << 14
+/**
+ * Adds an independent virtio-net device with the passt backend.
+ * Call to this function disables TSI backend.
```
Yes, the first one will be `eth0`, the second `eth1`, etc... I'll document this in both functions.

slp commented on include/libkrun.h:
```diff
@@ -265,7 +265,63 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+ * If no network devices are added, networking uses the TSI backend.
+ * This function should be called before krun_set_port_map.
+ *
+ * Returns:
+ *  Zero on success or a negative error number on failure.
+ */
+int32_t krun_add_net_passt(uint32_t ctx_id,
```
I like the idea, but I think we should leave this for libkrun-v2.x.x, as we don't this behavior for any other function in the API.

slp commented on include/libkrun.h:
```diff
@@ -265,7 +265,63 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+#define NET_GUEST_UFO 1 << 10
+#define NET_HOST_TSO4 1 << 11
+#define NET_HOST_TSO6 1 << 12
+#define NET_HOST_UFO 1 << 14
```
Added a comment referencing `uapi/linux/virtio_net.h`.

mtjhrc commented on include/libkrun.h:
```diff
@@ -283,6 +364,8 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
 int32_t krun_set_passt_fd(uint32_t ctx_id, int fd);
 
 /**
+ * DEPRECATED. Use krun_add_net_gvproxy instead.
```
You forgot to rename the reference to function in the doc:
```suggestion
 * DEPRECATED. Use krun_add_net_unixgram instead.
```

slp replied:
Ouch, fixed, thanks!

mtjhrc commented on include/libkrun.h:
```diff
@@ -265,7 +265,88 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+ * Returns:
+ *  Zero on success or a negative error number on failure.
+ */
+int32_t krun_add_net_passt(uint32_t ctx_id,
```
Since we are also renaming the `krun_add_gvproxy_path` to be more general, we should do that here too.
QEMU uses the paramter `-netdev stream` (https://www.qemu.org/docs/master/system/devices/net.html#to-use-socket-based-passt-interface). 

So maybe `krun_add_net_unixstream_fd()`? I don't know..., naming is hard, I don't care about the specific name, but for consistency we don't have to name it as `passt` specific.

nirs replied:
qemu -netdev socket also prefix the packet size to the packet. socket_vmnet is another application that can work with this format. It works with qemu -netdev socket and with lima vms, converting the prefix+payload to payload for apple virtualization datagram socket.

lima forwarding: https://github.com/lima-vm/lima/blob/73c5f9761bff67d9cb746e14a20a8d58bae34b39/pkg/vz/network_darwin.go#L67

If passt use the same format as qemu, than krunkit can also work with socket_vment, and we can use a generic name.

slp replied:
I've confirmed that our backend works with `socket_vmnet` too, so I went ahead and renamed the backend as `Unixstream`. I also extended the API to support pointing to the network proxy with either the path or a file descriptor. I think this should cover all our use cases in a reasonable way.

nirs commented on include/libkrun.h:
```diff
@@ -265,7 +265,107 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+ *              connection to the userspace network proxy.
+ *  "c_mac"   - MAC address as an array of 6 uint8_t entries.
+ *  "flags"   - generic flags for the network interface.
+ *  "vflags"  - virtio-net flags for the network interface.
```
2 flags are not great. We can have one, and make sure that our flags do no clash with virtio flags. We can use a mask to pick our flags or virtio flags. Based on the value above, we can have:

    #define NET_VIRTIO_FLAGS_MASK 0x0000ffff
    #define NET_INTERNAL_FLAGS_MASK 0xffff0000

mtjhrc replied:
Seems fine to me.
I'm not totally opposed to having single flags argument, but I'd say I prefer having two arguments like Sergio has done. The `vflags` change which features the guest sees are supported, wheres the `flags` are just flags that can also be used for arbitrary things to tweak ("quirks"), like the `VFKIT` magic header.

slp replied:
I also prefer having two separate arguments. It's easier to maintain and less error-prone.

nirs replied:
2 options is more error prone for users of the library. The main issue is
using NET_FLAG_VFKIT with flags, enabling NET_VFLAG_CSUM by mistake silently, or using NET_VFLAG_CSUM for flags, enabling vfkit_magic silently. The way to prevent this to make NET_FLAG_VFKIT unique, for example (1 << 16). With this we can detect if the flags are used incorectly.

But, if vflags and flags do not clash, there is no reason to use 2 variables and not need to check for incorrect usage. Users do not need to think which flags belongs to which variable and they cannot use the function in the wrong way.

The price for making usage of the library safer, masking flags before using them in the library.

nirs commented on include/libkrun.h:
```diff
@@ -265,7 +265,107 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+ *  "ctx_id"  - the configuration context ID.
+ *  "c_path"  - a null-terminated string representing the path
+ *              for the unixstream socket where the userspace
+ *              network proxy is listening.
```
Must be NULL if fd is not -1

slp replied:
That's nicer, added.

nirs commented on include/libkrun.h:
```diff
@@ -265,7 +265,107 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+                              uint8_t *const c_mac,
+                              uint32_t flags,
+                              uint32_t vflags);
+
```
Same issues as previous function.

Nice that the function are the same, if learn to use one you know how to use the other.

slp replied:
Also fixed for this one.

nirs commented on src/devices/src/virtio/net/device.rs:
```diff
@@ -65,8 +62,10 @@ unsafe impl ByteValued for VirtioNetConfig {}
+    UnixstreamFd(RawFd),
+    UnixstreamPath(PathBuf),
+    UnixgramFd(RawFd),
+    UnixgramPath(PathBuf, bool),
```
Nice

nirs commented on src/devices/src/virtio/net/unixgram.rs:
```diff
@@ -37,7 +42,9 @@ impl Gvproxy {
-        send(fd, &VFKIT_MAGIC, MsgFlags::empty()).map_err(ConnectError::SendingMagic)?;
+        if vfkit_magic {
+            send(fd, &VFKIT_MAGIC, MsgFlags::empty()).map_err(ConnectError::SendingMagic)?;
+        }
```
Nice. Maybe rename to send_vfkit_magic to make it more clear.

slp replied:
Renamed, thanks.

nirs commented on src/devices/src/virtio/net/unixgram.rs:
```diff
@@ -80,11 +87,11 @@ impl Gvproxy {
 
-impl NetBackend for Gvproxy {
+impl NetBackend for Unixgram {
     /// Try to read a frame from passt. If no bytes are available reports ReadError::NothingRead
```
We can replace the leftover "passt" with "peer"

slp replied:
I've replaced them with "network proxy" or "proxy".

nirs commented on src/devices/src/virtio/net/unixgram.rs:
```diff
@@ -126,7 +133,7 @@ impl NetBackend for Gvproxy {
     }
 
```
another leftover passt in line 111. We can also clean this up later to keep this change minimal, these leftovers were already there.

nirs commented on src/devices/src/virtio/net/unixstream.rs:
```diff
@@ -1,40 +1,77 @@
+        }
+
+        log::debug!(
+            "passt socket (fd {fd}) buffer sizes: SndBuf={:?} RcvBuf={:?}",
```
passt -> unixstream?

nirs commented on src/devices/src/virtio/net/unixstream.rs:
```diff
@@ -1,40 +1,77 @@
-            "passt socket (fd {passt_fd}) buffer sizes: SndBuf={:?} RcvBuf={:?}",
-            getsockopt(passt_fd, sockopt::SndBuf),
-            getsockopt(passt_fd, sockopt::RcvBuf)
+            "passt socket (fd {fd}) buffer sizes: SndBuf={:?} RcvBuf={:?}",
```
passt -> unixstream

nirs commented on src/libkrun/src/lib.rs:
```diff
@@ -752,8 +731,172 @@ pub unsafe extern "C" fn krun_set_data_disk(ctx_id: u32, c_disk_path: *const c_c
+    | NET_VFLAG_GUEST_UFO
+    | NET_VFLAG_HOST_TSO4
+    | NET_VFLAG_HOST_TSO6
+    | NET_VFLAG_HOST_UFO;
```
It is not clear what are the compat flags - maybe add a comment? Seems that the difference are NET_VFLAG_GUEST_TSO6 and NET_VFLAG_HOST_TSO6.

We can use this to make this more clear:

```rust
const NET_ALL_VFLAGS: u32 = NET_COMPAT_VFLAGS
    | NET_VFLAG_GUEST_TSO6
    | NET_VFLAG_HOST_TSO6;
```

slp replied:
Added a comment clarifying why that's there, thanks!

nirs commented on src/libkrun/src/lib.rs:
```diff
@@ -752,8 +731,172 @@ pub unsafe extern "C" fn krun_set_data_disk(ctx_id: u32, c_disk_path: *const c_c
+    };
+
+    if fd >= 0 && path.is_some() {
+        return -libc::EINVAL;
```
EINVAL without an error message is hard to work with. We have 4 different EINVAL errors for this function. How the user can tell what is the issue?

Can we have libkrun_error() to get the last error message? Probably not in the scope of this change, and I did not check the rest of the code.

mtjhrc replied:
> Can we have libkrun_error() to get the last error message? Probably not in the scope of this change, and I did not check the rest of the code. 

Yeah, so our error reporting is not great, but unfortunately we don't have anything like `libkrun_error()`, nor do we define our own error codes witch would help here. Currently the only way to see what went wrong for example with `krun_start_enter` is to check the log (which has to be enabled). Not great I know, but that is really something we need to change in v2 API more globally (otherwise it would get really confusing how errors work).

nirs commented on src/vmm/src/vmm_config/net.rs:
```diff
@@ -41,51 +44,28 @@ type Result<T> = result::Result<T, NetworkInterfaceError>;
         NetBuilder {
             // List of built network devices.
-            net_devices: Vec::new(),
+            list: VecDeque::new(),
```
Why do we need a deque? we only append to the end.

nirs reviewed (commented):
So much nicer with generic names!

nirs commented on include/libkrun.h:
```diff
@@ -265,7 +265,109 @@ int32_t krun_add_virtiofs2(uint32_t ctx_id,
+#define NET_FLAG_VFKIT 1 << 0
+
+/* Taken from uapi/linux/virtio_net.h */
+#define NET_VFLAG_CSUM 1 << 0
```
If we go with 2 flag arguments let's make the constants and the flags names more clear using "features" and "flags".

```
#define NET_FEATURE_CSUM 1<<0
...
 *  "features"  - virtio-net features for the network interface.
 *  "flags"     - generic flags for the network interface.
```

Switching the order because the virtio features are more important and flags are usually the last argument.

slp replied:
Sounds reasonable to me, I've just changed it.

slp commented:
@mtjhrc could you please take another look?

mtjhrc commented on examples/boot_efi.c:
```diff
@@ -23,6 +23,8 @@
 #define MAX_PATH 4096
 #endif
 
+#define COMPAT_NET_FLAGS NET_CSUM | NET_GUEST_CSUM | NET_GUEST_TSO4 | NET_GUEST_UFO | NET_HOST_TSO4 | NET_HOST_UFO
```
You forgot to update this to the new constant names

slp commented:
> Please fix the `boot_efi` example to not refer to non-existant functions/constants. Otherwise everything looks good.

Good catch, thanks. I've fixed that and also moved COMPAT_NET_FEATURES to `libkrun.h`, which seems like a better place for it.
